File: src/physdes/recti.py
from .interval import Interval
from .point import Point


class Rectangle(Point[Interval[int], Interval[int]]):
    def __init__(self, xcoord: Interval, ycoord: Interval):
        """[summary]

        Args:
            xcoord (Interval): [description]
            ycoord (Interval): [description]

        Examples:
            >>> a = Rectangle(Interval(3, 4), Interval(5, 6))
            >>> print(a)
            ([3, 4], [5, 6])
            >>> a3d = Rectangle(a, Interval(7, 8))  # Rectangle in 3d
            >>> print(a3d)
            (([3, 4], [5, 6]), [7, 8])
        """
        Point.__init__(self, xcoord, ycoord)

    @property
    def ll(self) -> Point[int, int]:
        """Lower left

        Returns:
            [type]: [description]

        Examples:
            >>> a = Rectangle(Interval(3, 4), Interval(5, 6))
            >>> print(a.ll)
            (3, 5)
        """
        return Point(self.xcoord.lb, self.ycoord.lb)

    @property
    def ur(self) -> Point[int, int]:
        """Upper right

        Returns:
            [type]: [description]

        Examples:
            >>> a = Rectangle(Interval(3, 4), Interval(5, 6))
            >>> print(a.ur)
            (4, 6)
        """
        return Point(self.xcoord.ub, self.ycoord.ub)

    # def copy(self):
    #     """[summary]

    #     Returns:
    #         [type]: [description]

    #     Examples:
    #         >>> a = Rectangle(Interval(3, 4), Interval(5, 6))
    #         >>> print(a.copy())
    #         ([3, 4], [5, 6])
    #         >>> a3d = Rectangle(a, Interval(7, 8))  # Rectangle in 3d
    #         >>> print(a3d.copy())
    #         (([3, 4], [5, 6]), [7, 8])
    #     """
    #     return Rectangle(self.xcoord, self.ycoord)

    # def __eq__(self, rhs) -> bool:
    #     return self.xcoord == rhs.xcoord and self.ycoord == rhs.ycoord

    def flip(self) -> "Rectangle":
        """[summary]

        Returns:
            [type]: [description]

        Examples:
            >>> a = Rectangle(Interval(3, 4), Interval(5, 6))
            >>> print(a.flip())
            ([5, 6], [3, 4])
            >>> a3d = Rectangle(a, Interval(7, 8))  # Rectangle in 3d
            >>> print(a3d.flip())
            ([7, 8], ([3, 4], [5, 6]))
        """
        return Rectangle(self.ycoord, self.xcoord)

    # `a` can be Point, VSegment, HSegment, or Rectangle
    def contains(self, other: Point) -> bool:
        """[summary]

        Args:
            obj (Point): [description]

        Returns:
            bool: [description]

        Examples:
            >>> a = Rectangle(Interval(30, 40), Interval(50, 60))
            >>> a.contains(Point(36, 53))
            True
            >>> a.contains(Rectangle(Interval(32, 38), Interval(51, 57)))
            True
            >>> a.contains(Rectangle(Interval(32, 38), Interval(51, 67)))
            False
        """
        return self.xcoord.contains(other.xcoord) and self.ycoord.contains(other.ycoord)

    def width(self) -> int:
        """[summary]

        Returns:
            [type]: [description]

        Examples:
            >>> a = Rectangle(Interval(30, 40), Interval(50, 62))
            >>> a.width()
            10
        """
        return self.xcoord.length()

    def height(self) -> int:
        """[summary]

        Returns:
            [type]: [description]

        Examples:
            >>> a = Rectangle(Interval(30, 40), Interval(50, 62))
            >>> a.height()
            12
        """
        return self.ycoord.length()

    def area(self) -> int:
        """[summary]

        Returns:
            [type]: [description]

        Examples:
            >>> a = Rectangle(Interval(30, 40), Interval(50, 62))
            >>> a.area()
            120
        """
        return self.xcoord.length() * self.ycoord.length()


class VSegment(Point[int, Interval[int]]):
    """
    Represents a VSegment.
    """

    def contains(self, other: Point) -> bool:
        """[summary]

        Args:
            a ([type]): [description]

        Returns:
            bool: [description]

        Examples:
            >>> a = VSegment(5, Interval(30, 40))
            >>> a.contains(Point(5, 33))
            True
            >>> a.contains(VSegment(5, Interval(33, 38)))
            True
            >>> a.contains(VSegment(6, Interval(33, 38)))
            False
        """
        return self.xcoord == other.xcoord and self.ycoord.contains(other.ycoord)

    def flip(self) -> "HSegment":
        """[summary]

        Returns:
            [type]: [description]

        Examples:
            >>> a = VSegment(5, Interval(30, 40))
            >>> print(a.flip())
            ([30, 40], 5)
        """
        return HSegment(self.ycoord, self.xcoord)


class HSegment(Point[Interval[int], int]):
    """
    Represents a HSegment.
    """

    # def __init__(self, xcoord, ycoord):
    #     """[summary]
    #
    #     Args:
    #         xcoord ([type]): [description]
    #         ycoord ([type]): [description]
    #
    #     Examples:
    #         >>> a = HSegment(Interval(3, 4), 5)
    #         >>> print(a)
    #         ([3, 4], 5)
    #         >>> a3d = HSegment(a, 7)  # HSegment in 3d
    #         >>> print(a3d)
    #         (([3, 4], 5), 7)
    #     """
    #     Point.__init__(self, xcoord, ycoord)

    # def copy(self):
    #     """[summary]

    #     Returns:
    #         [type]: [description]

    #     Examples:
    #         >>> a = HSegment(Interval(3, 4), 5)
    #         >>> print(a.copy())
    #         ([3, 4], 5)
    #         >>> a3d = HSegment(a, 7)  # HSegment in 3d
    #         >>> print(a3d.copy())
    #         (([3, 4], 5), 7)
    #     """
    #     return HSegment(self.xcoord, self.ycoord)

    # `a` can be Point or HSegment
    def contains(self, other) -> bool:
        """[summary]

        Args:
            a ([type]): [description]

        Returns:
            bool: [description]

        Examples:
            >>> a = HSegment(Interval(30, 40), 5)
            >>> a.contains(Point(33, 5))
            True
            >>> a.contains(HSegment(Interval(33, 38), 5))
            True
            >>> a.contains(HSegment(Interval(33, 38), 6))
            False
        """
        return self.ycoord == other.ycoord and self.xcoord.contains(other.xcoord)

    def flip(self) -> VSegment:
        """[summary]

        Returns:
            [type]: [description]

        Examples:
            >>> a = HSegment(Interval(30, 40), 5)
            >>> print(a.flip())
            (5, [30, 40])
        """
        return VSegment(self.ycoord, self.xcoord)

File: rust_ai/src/recti.rs
//! Rectangle and segment modules for geometric shapes
//!
//! This module provides types for rectangles and line segments in 2D space.

use std::fmt;
use num_traits::Num;
use approx::{AbsDiffEq, RelativeEq};

use crate::interval::Interval;
use crate::point::Point;
use crate::generic::{Overlaps, Contains, MinDistWith, Measure, Center};

/// An axis-aligned rectangle
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct Rectangle<T> {
    x_interval: Interval<T>,
    y_interval: Interval<T>,
}

impl<T> Rectangle<T> {
    /// Create a new rectangle with given x and y intervals
    pub fn new(x_interval: Interval<T>, y_interval: Interval<T>) -> Self {
        Self { x_interval, y_interval }
    }
    
    /// Get the x-interval
    pub fn x_interval(&self) -> &Interval<T> {
        &self.x_interval
    }
    
    /// Get the y-interval
    pub fn y_interval(&self) -> &Interval<T> {
        &self.y_interval
    }
    
    /// Get the lower-left point of the rectangle
    pub fn ll(&self) -> Point<T>
    where
        T: Copy,
    {
        Point::new(*self.x_interval.lb(), *self.y_interval.lb())
    }
    
    /// Get the upper-right point of the rectangle
    pub fn ur(&self) -> Point<T>
    where
        T: Copy,
    {
        Point::new(*self.x_interval.ub(), *self.y_interval.ub())
    }
    
    /// Get the width of the rectangle
    pub fn width(&self) -> T
    where
        T: Copy + Sub<Output = T>,
    {
        self.x_interval.width()
    }
    
    /// Get the height of the rectangle
    pub fn height(&self) -> T
    where
        T: Copy + Sub<Output = T>,
    {
        self.y_interval.width()
    }
    
    /// Get the area of the rectangle
    pub fn area(&self) -> T
    where
        T: Copy + Sub<Output = T> + Mul<Output = T>,
    {
        self.width() * self.height()
    }
    
    /// Flip the rectangle (swap x and y coordinates)
    pub fn flip(&self) -> Self
    where
        T: Copy,
    {
        Self::new(self.y_interval, self.x_interval)
    }
}

impl<T: fmt::Display> fmt::Display for Rectangle<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "({}, {})", self.x_interval, self.y_interval)
    }
}

// Implement generic traits for Rectangle
impl<T> Overlaps for Rectangle<T>
where
    T: PartialOrd + Copy,
{
    fn overlaps(&self, other: &Self) -> bool {
        self.x_interval.overlaps(&other.x_interval) && 
        self.y_interval.overlaps(&other.y_interval)
    }
}

impl<T> Contains for Rectangle<T>
where
    T: PartialOrd + Copy,
{
    fn contains(&self, other: &Self) -> bool {
        self.x_interval.contains(&other.x_interval) && 
        self.y_interval.contains(&other.y_interval)
    }
}

impl<T> Contains<Point<T>> for Rectangle<T>
where
    T: PartialOrd + Copy,
{
    fn contains(&self, other: &Point<T>) -> bool {
        self.x_interval.contains(other.x()) && 
        self.y_interval.contains(other.y())
    }
}

impl<T> MinDistWith for Rectangle<T>
where
    T: Copy + PartialOrd + Sub<Output = T> + Signed,
{
    type Output = T;
    
    fn min_dist_with(&self, other: &Self) -> T {
        let dx = self.x_interval.min_dist_with(&other.x_interval);
        let dy = self.y_interval.min_dist_with(&other.y_interval);
        dx + dy  // Manhattan distance
    }
}

impl<T> MinDistWith<Point<T>> for Rectangle<T>
where
    T: Copy + PartialOrd + Sub<Output = T> + Signed,
{
    type Output = T;
    
    fn min_dist_with(&self, other: &Point<T>) -> T {
        let dx = self.x_interval.min_dist_with(other.x());
        let dy = self.y_interval.min_dist_with(other.y());
        dx + dy  // Manhattan distance
    }
}

impl<T> Measure for Rectangle<T>
where
    T: Copy + Sub<Output = T> + Mul<Output = T>,
{
    type Output = T;
    
    fn measure(&self) -> T {
        self.area()
    }
}

impl<T> Center for Rectangle<T>
where
    T: Copy + Add<Output = T> + Sub<Output = T> + Num + From<i32>,
{
    type Output = Point<T>;
    
    fn center(&self) -> Point<T> {
        Point::new(self.x_interval.center(), self.y_interval.center())
    }
}

/// A vertical line segment
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct VSegment<T> {
    x: T,
    y_interval: Interval<T>,
}

impl<T> VSegment<T> {
    /// Create a new vertical segment
    pub fn new(x: T, y_interval: Interval<T>) -> Self {
        Self { x, y_interval }
    }
    
    /// Get the x-coordinate
    pub fn x(&self) -> &T {
        &self.x
    }
    
    /// Get the y-interval
    pub fn y_interval(&self) -> &Interval<T> {
        &self.y_interval
    }
    
    /// Flip the segment (vertical to horizontal)
    pub fn flip(&self) -> HSegment<T>
    where
        T: Copy,
    {
        HSegment::new(self.y_interval, self.x)
    }
}

impl<T: fmt::Display> fmt::Display for VSegment<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "VSegment(x={}, y={})", self.x, self.y_interval)
    }
}

/// A horizontal line segment
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub struct HSegment<T> {
    x_interval: Interval<T>,
    y: T,
}

impl<T> HSegment<T> {
    /// Create a new horizontal segment
    pub fn new(x_interval: Interval<T>, y: T) -> Self {
        Self { x_interval, y }
    }
    
    /// Get the x-interval
    pub fn x_interval(&self) -> &Interval<T> {
        &self.x_interval
    }
    
    /// Get the y-coordinate
    pub fn y(&self) -> &T {
        &self.y
    }
    
    /// Flip the segment (horizontal to vertical)
    pub fn flip(&self) -> VSegment<T>
    where
        T: Copy,
    {
        VSegment::new(self.y, self.x_interval)
    }
}

impl<T: fmt::Display> fmt::Display for HSegment<T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "HSegment(x={}, y={})", self.x_interval, self.y)
    }
}

// Implement generic traits for segments
impl<T> Overlaps for VSegment<T>
where
    T: PartialOrd + Copy,
{
    fn overlaps(&self, other: &Self) -> bool {
        self.x == other.x && self.y_interval.overlaps(&other.y_interval)
    }
}

impl<T> Contains for VSegment<T>
where
    T: PartialOrd + Copy,
{
    fn contains(&self, other: &Self) -> bool {
        self.x == other.x && self.y_interval.contains(&other.y_interval)
    }
}

impl<T> Contains<Point<T>> for VSegment<T>
where
    T: PartialOrd + Copy,
{
    fn contains(&self, other: &Point<T>) -> bool {
        self.x == *other.x() && self.y_interval.contains(other.y())
    }
}

impl<T> Overlaps for HSegment<T>
where
    T: PartialOrd + Copy,
{
    fn overlaps(&self, other: &Self) -> bool {
        self.y == other.y && self.x_interval.overlaps(&other.x_interval)
    }
}

impl<T> Contains for HSegment<T>
where
    T: PartialOrd + Copy,
{
    fn contains(&self, other: &Self) -> bool {
        self.y == other.y && self.x_interval.contains(&other.x_interval)
    }
}

impl<T> Contains<Point<T>> for HSegment<T>
where
    T: PartialOrd + Copy,
{
    fn contains(&self, other: &Point<T>) -> bool {
        self.y == *other.y() && self.x_interval.contains(other.x())
    }
}

// Implement approximate equality
impl<T> AbsDiffEq for Rectangle<T>
where
    T: AbsDiffEq<Epsilon = T> + Copy,
    T::Epsilon: Copy,
{
    type Epsilon = T::Epsilon;
    
    fn default_epsilon() -> Self::Epsilon {
        T::default_epsilon()
    }
    
    fn abs_diff_eq(&self, other: &Self, epsilon: Self::Epsilon) -> bool {
        self.x_interval.abs_diff_eq(&other.x_interval, epsilon) &&
        self.y_interval.abs_diff_eq(&other.y_interval, epsilon)
    }
}

impl<T> RelativeEq for Rectangle<T>
where
    T: RelativeEq<Epsilon = T> + Copy,
    T::Epsilon: Copy,
{
    fn default_max_relative() -> Self::Epsilon {
        T::default_max_relative()
    }
    
    fn relative_eq(&self, other: &Self, epsilon: Self::Epsilon, max_relative: Self::Epsilon) -> bool {
        self.x_interval.relative_eq(&other.x_interval, epsilon, max_relative) &&
        self.y_interval.relative_eq(&other.y_interval, epsilon, max_relative)
    }
}

// Need to import std::ops for the generic constraints
use std::ops::{Add, Sub, Mul};
use num_traits::Signed;